csharp_dotnetcore/21.Custom-Dialogs/CustomDialogBot.cs

```diff
@@ -0,0 +1,139 @@
+    /// each time they're requested. For each Activity received, a new instance of this
+    /// class is created. Objects that are expensive to construct, or have a lifetime
+    /// beyond the single turn, should be carefully managed.
+    /// For example, the <see cref="MemoryStorage"/> object and associated
+    /// <see cref="IStatePropertyAccessor{T}"/> object are created with a singleton lifetime.
+    /// </summary>
+    /// <seealso cref="https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-2.1"/>
+    public class CustomDialogBot : IBot
+    {
+        private readonly BotAccessors _accessors;
+
+        /// <summary>
+        /// The <see cref="DialogSet"/> that contains all the Dialogs that can be used at runtime.
+        /// </summary>
+        private DialogSet _dialogs;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CustomDialogBot"/> class.
+        /// </summary>
+        /// <param name="accessors">A class containing <see cref="IStatePropertyAccessor{T}"/> used to manage state.</param>
+        public CustomDialogBot(BotAccessors accessors)
+        {
+            _accessors = accessors ?? throw new ArgumentNullException(nameof(accessors));
+
+            // The DialogSet needs a DialogState accessor, it will call it when it has a turn context.
+            _dialogs = new DialogSet(accessors.ConversationDialogState);
+
+            var name_slots = new List<SlotDetails>
+            {
+                new SlotDetails { Name = "first", PromptId = "text", PromptText = "Please enter your first name." },
+                new SlotDetails { Name = "last", PromptId = "text", PromptText = "Please enter your last name." },
+            };
+
+            var address_slots = new List<SlotDetails>
+            {
+                new SlotDetails { Name = "street", PromptId = "text", PromptText = "Please enter the street." },
+                new SlotDetails { Name = "city", PromptId = "text", PromptText = "Please enter the city." },
+                new SlotDetails { Name = "zip", PromptId = "text", PromptText = "Please enter the zip." },
+            };
+
+            var slots = new List<SlotDetails>
+            {
+                new SlotDetails { Name = "fullname", PromptId = "fullname" },
+                new SlotDetails { Name = "age", PromptId = "number", PromptText = "Please enter your age." },
+                new SlotDetails { Name = "shoesize", PromptId = "number", PromptText = "Please enter your shoe size." },
+                new SlotDetails { Name = "nickname", PromptId = "text", PromptText = "Please enter your nickname." },
+                new SlotDetails { Name = "address", PromptId = "address" },
+            };
+
+            // Add the various dialogs we will be using to teh DialogSet
+            _dialogs.Add(new SlotFillingDialog("address", address_slots));
+            _dialogs.Add(new SlotFillingDialog("fullname", name_slots));
```

Comment:
It seems overkill to, on EVERY TURN, allocate all of these dialogs and then not use most of them. 

Is there a way to improve this?